### PR TITLE
Place utils contracts in remapped directories when appropriate

### LINKED ISSUE #160: `ParserError` file not found on truffle-compilation of scribble instrumented code
If you have a truffle repo with 

1) `invariant` annotation on a contract X, which inherits from contract Y, where Y is defined in a package under `node_modules/`
2) And you instrument and arm the code
3) Try to run `truffle compile`

You will run into compilation failures like this one:

```
ParserError: Source "contracts/__scribble_ReentrancyUtils.sol" not found
 --> @openzeppelin/contracts/access/AccessControlEnumerable.sol:8:1:
  |
8 | import "./../../../../contracts/__scribble_ReentrancyUtils.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The issue is that our instrumentation needs to modify all base contracts, including the ones under `node_modules/`. As part of the modification, they need to inherit from the `__scribble_ReentrancyUtils` contract, which is usually placed at the root of the `contracts/` directory. This adds an import from the `node_modules/` directory back into the `contracts/` directory.

In other development environments, to compile such contracts users need to specify a path remapping like `@openzeppelin=node_modules/@openzeppelin`. As a result, the solc compiler will preform the following steps when resolving the import declaration above:

1. Append the relative import path to the current file directory path. In the above example we would append `./../../../../contracts/__scribble_ReentrancyUtils.sol` to `@openzeppelin/contracts/access/` to obtain
`@openzeppelin/contracts/access/./../../../../contracts/__scribble_ReentrancyUtils.sol`
2. Apply any path remappings. In the above example we would apply  `@openzeppelin=node_modules/@openzeppelin` to `@openzeppelin/contracts/access/./../../../../contracts/__scribble_ReentrancyUtils.sol` to obtain
`node_modules/@openzeppelin/contracts/access/./../../../../contracts/__scribble_ReentrancyUtils.sol`
3. Normalize the path, collapsing any . and .. In the above example this would transform `node_modules/@openzeppelin/contracts/access/./../../../../contracts/__scribble_ReentrancyUtils.sol` into `contracts/__scribble_ReentrancyUtils.sol`
4. Resolve the obtained path.

However in the case of truffle, it seems to be doing its own separate import path resolution, and building the complete standard json input to pass in to the compiler (around [here](https://github.com/trufflesuite/truffle/tree/develop/packages/resolver)). That doesn't seem to like our relative imports.

Steps to reproduce:

```
git clone git@github.com:cd1m0/scribble-truffle-remapping-bug.git
cd scribble-truffle-remapping-bug/
npm install
scribble contracts/Foo.sol --output-mode files --arm --path-remapping '@openzeppelin=node_modules/@openzeppelin' --compiler-version 0.8.11
truffle compile
```

You should see the following errors:

```
ParserError: Source "contracts/__scribble_ReentrancyUtils.sol" not found
 --> @openzeppelin/contracts/token/ERC20/ERC20.sol:8:1:
  |
8 | import "./../../../../../contracts/__scribble_ReentrancyUtils.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

,ParserError: Source "contracts/__scribble_ReentrancyUtils.sol" not found
 --> @openzeppelin/contracts/utils/Context.sol:5:1:
  |
5 | import "./../../../../contracts/__scribble_ReentrancyUtils.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Compilation failed. See above.
```

### PR DESCRIPTION
**Pull request by cd1m0**
This PR: 

* Place utils contracts under `node_modules` when package remapping is present (Fixes #160)
* Add support for function lists in using for directives (added in solc 0.8.13)

### CODE REVIEW

**Review by blitz-1306 — COMMENTED** [review]
Great job.

There are more issues with Solidity 0.8.13 and I'm not sure whether we should fix tem as part of this PR or a followup one. Lets assume we have following setup:
- **sample.sol**
    ```solidity
    pragma solidity ^0.8.13;

    type X is uint256;

    function add(X a, X b) pure returns (X) {
        return X.wrap(X.unwrap(a) + X.unwrap(b));
    }

    using { add } for X global;
    ```
- **other.sol**
    ```solidity
    pragma solidity ^0.8.13;

    import { X } from "./sample.sol";

    function add(X a, X b) pure returns (uint256) {
        return 1;
    }

    contract Test {
        function main() public returns (X) {
            X a = X.wrap(1);
            X b = X.wrap(2);

            return a.add(b);
        }
    }
    ```

Then run command:
```bash
scribble other.sol --instrumentation-metadata-file instr.json
```

There are multiple issues with that case:
- The are typechecking (`typecheck.ts`) issues with the bound methods.
- The `merge()` does not behave properly for the valid `Identifier`s in imports.
- The flattening also does not properly respect two different free functions with the same name `add(...)`.
- I also suspect that flattening will not correctly handle user-defined value types with the same name, but did not composed a sample to verify that this may happen.

I have a few fixes, so I'm wondering if I should push them to current branch or create other one? Or, would you like to investigate personally? (:

File: package.json [hunk 1]
```diff
@@ -18,18 +18,19 @@
         "fs-extra": "^10.0.1",
         "logplease": "^1.2.15",
         "semver": "^7.3.5",
-        "solc-typed-ast": "^9.0.0",
+        "solc-typed-ast": "^9.1.1",
         "src-location": "^1.1.0",
         "findup-sync": "^5.0.0",
-        "yaml": "^1.10.2"
+        "yaml": "^1.10.2",
+        "glob": "^7.2.0"
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply package.json:25]: Fixed in [d51c8cc](https://github.com/ConsenSys/scribble/pull/161/commits/d51c8cc0a7decc7a06ef121cd1e42bbaf3d62342)

File: src/bin/scribble.ts [hunk 2]
```diff
@@ -440,8 +443,71 @@ function writeOut(contents: string, fileName: string) {
...
+
+/**
+ * This functions picks the location where we will place the `__scribble_ReentrancyUtils.sol` file.
+ * Optionally, in some cases it also returns the exact import path we should use. There are 3 cases:
+ *
+ * 1. If the user specified a --utils-output-path command line option just use that
+ * 2. If there is some path remapping from say @openzeppelin to node_modules/@openzeppelin and there is a package.json
+ *    somewhere inside in say node_modules/@openzeppelin/contracts then put the file in the same directory
+ *    as the package.json. Also return an import path starting with the remap prefix (e.g. @openzeppelin)
+ * 3. Otherwise put the utils file in the same director as the first target contract.
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/bin/scribble.ts]: Fixed in [d51c8cc](https://github.com/ConsenSys/scribble/pull/161/commits/d51c8cc0a7decc7a06ef121cd1e42bbaf3d62342)

File: src/spec-lang/tc/typecheck.ts [hunk 3]
```diff
@@ -1733,20 +1734,37 @@ export function tcMemberAccess(expr: SMemberAccess, ctx: STypingCtx, typeEnv: Ty
...
+            usingForInScope.push(...base.vUsingForDirectives);
+        }
+
+        usingForInScope.push(...contract.vScope.vUsingForDirectives);
+
+        for (const usingFor of usingForInScope) {
+            const usingForApplies =
+                usingFor.vTypeName === undefined
+                    ? true
+                    : eq(typeNameToTypeNode(usingFor.vTypeName), generalBaseT);
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/spec-lang/tc/typecheck.ts]: Fixed in [d51c8cc](https://github.com/ConsenSys/scribble/pull/161/commits/d51c8cc0a7decc7a06ef121cd1e42bbaf3d62342)

**Comment by cd1m0** [comment]
Given that we have user complaints about #160 I would rather handle the remaining issues with 0.8.13 as a separate PR. I would even be down to do a patch release after we merge this and then fix the remaining 0.8.13 issues in the next patch. Would you be ok with this idea?

**Review by blitz-1306 — APPROVED** [review]

**Pull request merged**